### PR TITLE
docs: Use jupyterlite-sphinx to embed JupyterLite in docs

### LINKED ISSUE #2136: Building Jupyterlite for docs rather than using a demo site
> FYI, I'd probably recommend building Jupyterlite yourself, rather than using a demo site that is not stable. Something like https://github.com/henryiii/se-for-sci/blob/b97673d82361f1be50c1519efac2a1de50e1e157/noxfile.py#L8-L15, for example.

_Originally posted by @henryiii in https://github.com/scikit-hep/pyhf/issues/2132#issuecomment-1464638354_

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #2136

* Add `jupyterlite-sphinx`, `jupyterlite-pyodide-kernel`, and `jupytext` to 'docs' extra.
* Add `docs/lite/jupyterlite.py` as `jupytext` 'py:percent' textfile representation of `jupyterlite` demo notebook.
* Use `jupyterlite-sphinx` extension to embed JupyterLite in the docs.
   - Use `jupytext` to convert the `jupyterlite.py` Python file into a Jupyter notebook at docs build time.
   - Use the [`retrolite` directive](https://jupyterlite-sphinx.readthedocs.io/en/latest/directives/retrolite.html) to embed the [RetroLab Lite JupyterLab distribution](https://jupyterlite.readthedocs.io/en/latest/_static/retro/notebooks/index.html?path=intro-retro.ipynb) in the docs and load the converted notebook at launch time. This also nicely requires the user to request installing and loading as opposed to doing so on page load.
   - Use the `jupyterlite` `memoryStorageDriver` to disable local storage to avoid users having to manually clear their browser cache to reset to a default state.
* Remove `docs/generate_jupyterlite_iframe.py` in favor of embedding JupyterLite.
* Add ignores for pre-commit hooks for the `docs/lite/jupyterlite.py` and add check in docs building that the built html/lite directory is not empty.

Users will now see this on the landing page:

before-run

which expands to the following if they click:

after-run

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add `jupyterlite-sphinx`, `jupyterlite-pyodide-kernel`, and `jupytext` to 'docs' extra.
* Add docs/lite/jupyterlite.py as jupytext 'py:percent' textfile representation
  of jupyterlite demo notebook.
* Use jupyterlite-sphinx extension to embed JupyterLite in the docs.
   - Use jupytext to convert the jupyterlite.py Python file into a Jupyter notebook
     at docs build time.
   - Use the retrolite directive to embed the RetroLab Lite JupyterLab distribution in
     the docs and load the converted notebook at launch time. This also nicely requires
     the user to request installing and loading as opposed to doing so on page load.
   - Use the jupyterlite memoryStorageDriver to disable local storage to avoid users
     having to manually clear their browser cache to reset to a default state.
* Remove docs/generate_jupyterlite_iframe.py in favor of embedding JupyterLite.
* Add ignores for pre-commit hooks for the docs/lite/jupyterlite.py and add check
  in docs building that the built html/lite directory is not empty.

Co-authored-by: Angus Hollands <goosey15@gmail.com>
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@agoose77 I'm following the work that you've done in Awkward to get the ["Try it" page](https://awkward-array.org/doc/main/getting-started/try-awkward-array.html) working

https://github.com/scikit-hep/awkward/blob/66b620958b9e0bd2f16787b66596a22249e17336/docs/getting-started/try-awkward-array.md?plain=1#L3

so I might but you on this later.

**Comment by matthewfeickert** [comment]
Relevant RTD build: https://pyhf--2187.org.readthedocs.build/en/2187/

File: docs/jupyterlite.rst [hunk 1]
```diff
@@ -1,30 +1,29 @@
...
-      width="100%"
-      height="500px"
-   ></iframe>
+  Comment: Use https://github.com/jupyterlite/jupyterlite-sphinx
+
+.. replite::
+   :kernel: python
+   :height: 600px
+   :prompt: Try pyhf!
+   :prompt_color: #dc3545
```
Comment (matthewfeickert) [inline docs/jupyterlite.rst:19]: Should make this a pyhf blue.

File: docs/jupyterlite.rst [hunk 2]
```diff
@@ -1,30 +1,29 @@
 Try out now with JupyterLite_
 -----------------------------
 
-.. admonition:: To try pyhf code in the browser with the Pyolite JupyterLite kernel:
+.. admonition:: Click below to try pyhf in the browser:
     :class: dropdown
 
      #. Type (or copy and paste) code in the input cell.
      #. To execute the code, press ``Shift + Enter`` or click on the (Run) ▶ button in the toolbar.
```
Comment (matthewfeickert) [inline docs/jupyterlite.rst:8]: At the moment with `replite` there is no toolbar anymore. Might want to work on flipping over to `jupyterlite` explicitly.

Comment (agoose77) [reply docs/jupyterlite.rst:8]: You probably want `retrolite`, which gives a single notebook (that's what we use on awkward-array.org)

**Comment by matthewfeickert** [comment]
For reasons I'm not clear on, for some reason the in-memory filesystem created by Emscripten is caching an old version of the notebook that was saved. I'm not sure how to clear it.

cached

File: docs/Makefile [hunk 3]
```diff
@@ -54,6 +54,7 @@ clean:
 
 .PHONY: html
 html:
+	jupytext --to ipynb lite/jupyterlite.py
```
Comment (matthewfeickert) [inline docs/Makefile:57]: Need to move this into the actual Sphinx build to get this to work in RTD.

**Comment by agoose77** [comment]
> For reasons I'm not clear on, for some reason the in-memory filesystem created by Emscripten is caching an old version of the notebook that was saved. I'm not sure how to clear it.
> 
> cached

Yes, this is a pain point. By default, you have to clear your browser cache to erase it.

You can enable the memory storage driver to fix this: https://github.com/scikit-hep/awkward/blob/main/docs/lite/jupyter-lite.json

**Comment by matthewfeickert** [comment]
> Yes, this is a pain point. By default, you have to clear your browser cache to erase it.
> 
> You can enable the memory storage driver to fix this: https://github.com/scikit-hep/awkward/blob/main/docs/lite/jupyter-lite.json

Thanks @agoose77, that's really helpful!

I think the answer is "no, that's now how the Emscripten filesystem works" but is there a way (using jupyterlite not retrolite, so as to get a file browser) to make the inmemory files visible to a browser so that one could do something like make a figure in matplotlib and then save the figure but be able to see the saved file that exists in the Emscripten filesystem in the browser to download it to a local filesystem?


<details>
<summary>Example code:</summary>

```python
import piplite
await piplite.install(["pyhf==0.7.1", "matplotlib>=3.0.0"])
%matplotlib inline

import pyhf
import numpy as np
import matplotlib.pyplot as plt
from pyhf.contrib.viz import brazil

pyhf.set_backend("numpy")
model = pyhf.simplemodels.uncorrelated_background(
    signal=[10.0], bkg=[50.0], bkg_uncertainty=[7.0]
)
data = [55.0] + model.config.auxdata

poi_vals = np.linspace(0, 5, 41)
results = [
    pyhf.infer.hypotest(
        test_poi, data, model, test_stat="qtilde", return_expected_set=True
    )
    for test_poi in poi_vals
]

fig, ax = plt.subplots()
fig.set_size_inches(7, 5)
brazil.plot_results(poi_vals, results, ax=ax);

fig.savefig("example.png")  # Want to be able to see this file in browser and then download it.
```

</details>

If I were to add something like

```python
...
jupyterlite_contents = ["lite/*"]
```

to `docs/conf.py` then I can see all the files that exist there in the JupyterLite file browser (as seen in the attached image) and I'm able to download them to my local filesystem. However, I don't seem to be able to write a file to this path.

path-in-memory

File: docs/conf.py [hunk 4]
```diff
@@ -15,12 +15,18 @@
...
 import sys
+from pathlib import Path
+
 from pkg_resources import get_distribution
 
 sys.path.insert(0, str(Path('./exts').resolve()))
 
+# Convert jupyterlite example to ipynb
+docs_dir = Path(__file__).resolve().parent
+subprocess.run(["jupytext", "--to", "ipynb", docs_dir / "lite" / "jupyterlite.py"])
```
Comment (matthewfeickert) [inline docs/conf.py:28]: I've opened up https://github.com/mwouts/jupytext/issues/1066 to see if I can avoid this easily.

**Comment by matthewfeickert** [comment]
I'm going to approve and merge this myself. As always, PRs approved by a single core dev can be reverted as needed by the rest of the dev team.

I'll also note that any reviews after merge are still welcome (and feel free to open up follow up Issues).

@alexander-held this also takes care of the problems you pointed out with automatically loading things on page visit being bad for people on limited network connections.

**Pull request merged**